### DISCUSSION TITLE
Is this Unresolved reference 'insert' on the String struct because of SKIP?

### DISCUSSION THREAD

**Discussion by davidakoontz** [skiptools/skip#195] [Q&A] [answered]
![Screenshot 2024-08-15 at 5 18 02 PM](https://github.com/user-attachments/assets/4176ba7b-a4a8-474a-be76-a730fa0656bf)

I'm pulling my hair out - trying to understand why the Compile is giving me all these:
**Unresolved reference 'insert'.**  Errors when trying to use the simple plan old  String.insert() function.

Is it because SKIP does NOT support this API call yet?

Looking at the Skip-Lib String - I do NOT see 'insert' but do I interpret that as NOT Supported?  Unsure!
<img width="1370" alt="Screenshot 2024-08-15 at 5 24 16 PM" src="https://github.com/user-attachments/assets/ba585aac-8ee8-4c16-935c-f0d306d36320">

**Comment by marcprux** [comment, marked as answer]
As you point out in the reference to https://skip.tools/docs/modules/skip-lib/#swift-standard-library-support, Kotlin strings are not mutable, and so we cannot support functions like `insert`. The workaround is to do something like:

```
greet = atSign + greet
```

**Comment by davidakoontz** [comment]
Hey Marc, @marcprux ,  as I've pondered this issue a while now, it is not making sense to me.

Swift Strings are also non-mutable and yet the Swift Lib allows many functions that mutate the original String.  I believe the Kotlin library does a very similar trick.  They return a *new* string upon "write".  This would be very similar to how the capitalized computed property works.  And how the function trimmingCharacters(in set: CharacterSet) -> String works.

So since I've got an App that is going to need a few String mutating functions...  I guess I have to write them myself.  OK... is there a way I could contribute this code to the SKIP project... if it was of high quality/robustness?

Do you have ANY suggestion for how to start ....

**Comment by marcprux** [comment]
Functions like `trimmingCharacters(in set: CharacterSet) -> String` return a new String, rather than mutating the existing string. We also do that in Skip. For example, `String.lowercased()` works fine, because it is returning a new String. It is the mutating functions on built-in types that we are unable to support. For example, we can't support `booleanField.toggle()` for this reason – the workaround is similar: do `booleanField = !booleanField` instead.

**Reply by davidakoontz** [reply]
Ahhh, thank you @marcprux;  now I'm starting to get the lay of the land...  it's Swift's **mutating func** that is not supported / compatible with Kotlin.  OK I'm slow but starting to catch the drift.  I know next to nothing about Kotlin world.

So the acceptable solution with Swift mutating func pattern is ... what?

```
var phrase = "Backya!"
let position = 4
#if !SKIP 
    phrase.insert("@", at: position)   // "Back@ya!"
#else
     // kotlin - workaround for string.insert
     // written in Swift code - but transpired to Kotlin
     let wordOneEndPosition = phrase.firstIndex(of: "y")!
     let wordOne = phrase[...wordOneEndPosition]
     let sentence = wordOne + "@" + "ya!"
     // now do something with sentence
     let phrase = sentence
#end if
    
```

Is this correct thinking?

**Comment by davidakoontz** [comment]
Still seeking understanding @marcprux,

Docs said: "The following table summarizes SkipLib’s Swift Standard Library API support on Android. Anything **not listed** here is likely not supported. Note that in your iOS-only code - i.e. code within #if !SKIP blocks - you can use any API you want."


I see that in **Skip-Lib the Collection** contains:
     **_mutating func insert(_ newElement: Element, at i: Int)_**
     
And does String implement Collection as a protocol?  Or maybe NOT???

Should I be looking to cast my String to a Collection?  So that I can use the **insert** method?
